### PR TITLE
Makefile update

### PR DESCRIPTION
**Pull request by danmois**
Implementation of first makefile features regarding the use of a specific SDK.

### CODE REVIEW

**Review by amiclaus — COMMENTED** [review]
It looks like these changes broke the `Build Projects` pipeline.
If you add `staging/` prefix to your branch: https://github.com/analogdevicesinc/no-OS/blob/f2c0ffea809196425300d2cdc39174959c54493f/build_projects.yml#L4
 the builds will be triggered on your custom branch, prior to doing a PR so you can debug the issues easier and double check that everything works fine :).

File: tools/scripts/build_projects.py [hunk 1]
```diff
@@ -192,11 +192,11 @@ def build(self):
...
-			err = run_cmd(cmd + ' clean_all')
+			err = run_cmd(cmd + ' reset')
 			if err != 0:
 				return err
 
-		err = run_cmd(cmd + ' update_srcs')
+		err = run_cmd(cmd + ' update')
 		if err != 0:
 			return err
 		err = run_cmd(cmd + ' -j%d re' % (multiprocessing.cpu_count() - 1))
```
Comment (amiclaus) [inline tools/scripts/build_projects.py:202]: `make re` was removed with d71c363baa61b574042e58f3a037ef32d350c02e

File: tools/scripts/generic.mk [hunk 2]
```diff
@@ -379,18 +380,11 @@ clean_all: clean_libs
 	@$(call print,[Delete] $(BUILD_DIR))
 	$(MUTE) $(call remove_dir,$(BUILD_DIR)) $(HIDE)
 
-PHONY += ca
```
Comment (buha) [inline tools/scripts/generic.mk:382]: I would rephrase the commit title to "...: remove obsolete/unneeded rules" or something like that

File: tools/scripts/build_projects.py (hunk 1 above)
Comment (buha) [reply tools/scripts/build_projects.py:202]: yes, true, we need to make sure all the changes in this PR don't break our CI builds

File: tools/scripts/generic.mk [hunk 3]
```diff
@@ -366,7 +367,10 @@ standalone:
 	$(MUTE) $(MAKE) --no-print-directory project LINK_SRCS=n MAKEFLAGS=$(MAKEOVERRIDES)
 
 # Build project using SDK
-project_build: $(PLATFORM)_project_build
+sdkbuild: $(PLATFORM)_project_build
```
Comment (buha) [inline tools/scripts/generic.mk:370]: Let's also rename $(PLATFORM)_project_build ---> $(PLATFORM)_sdkbuild so that it's easier to follow the script. This is internal-only, users wouldn't care, but it makes these scripts more maintainable... same for $(PLATFORM)_project_clean ---> $(PLATFORM)_sdkclean

**Review by buha — COMMENTED** [review]
I also think we should do the changes stm32.mk and aducm3029.mk platforms in this same PR, define the rules $(PLATFORM)_sdkbuild, $(PLATFORM)_sdkclean, $(PLATFORM)_sdkopen in all of them.
If we only do it for xilinx it will fail when you `make sdkbuild` on any other platform.

**Comment by danmois** [comment]
V2

- Removed also the ra rule

File: tools/scripts/generic.mk (hunk 2 above)
Comment (danmois) [reply tools/scripts/generic.mk:382]: Done

File: tools/scripts/build_projects.py (hunk 1 above)
Comment (danmois) [reply tools/scripts/build_projects.py]: Done

**Comment by danmois** [comment]
V3

- renamed $(PLATFORM)_project_build to $(PLATFORM)_sdkbuild
- renamed $(PLATFORM)_project_clean to $(PLATFORM)_sdkclean

File: tools/scripts/generic.mk (hunk 3 above)
Comment (danmois) [reply tools/scripts/generic.mk]: Done

File: tools/scripts/generic.mk [hunk 4]
```diff
@@ -375,8 +375,8 @@ clean:
 	$(MUTE) $(call remove_file,$(BINARY) $(OBJS) $(ASM_OBJS)) $(HIDE)
 
 # Remove the whole build directory
-PHONY += clean_all
-clean_all: clean_libs
+PHONY += reset
```
Comment (adrimbarean) [inline tools/scripts/generic.mk:382]: Is "reset" the correct name here? To me "reset" implies that after the clean operation is done the build starts again. Would "purge" be a better choice?

File: tools/scripts/platform/xilinx/build_project.tcl [hunk 5]
```diff
@@ -5,6 +5,6 @@ if { $argc != 1 } {
 
 sdk setws [lindex $argv 0]
 
-projects -build -name app -type app
+app build -name app
```
Comment (adrimbarean) [inline tools/scripts/platform/xilinx/build_project.tcl:8]: I would either make a visible note about this change in the commit description or place it in a separate commit. It probably is best if placed in a separate commit because, if I'm not mistaken, this it an update to the make process following an update to the Xilinx tools. I'd say it should special notice.

**Review by adrimbarean — COMMENTED** [review]
A couple of questions.

File: tools/scripts/generic.mk (hunk 4 above)
Comment (danmois) [reply tools/scripts/generic.mk:382]: I would say that "purge" is a term that is usual for the removal of sensitive data, while "reset" implies the return to the initial state. However, I would be happy to organize a poll on this matter, and choose the proper naming.

File: tools/scripts/platform/xilinx/build_project.tcl (hunk 5 above)
Comment (danmois) [reply tools/scripts/platform/xilinx/build_project.tcl:8]: Yes, this is true. It is the current recommended (by the Xilinx tools) way of dealing with projects.

**Comment by danmois** [comment]
V4

- addition of the sdkclean rule for the STM32 platform
- correction of echo message of remove file function in genetic.mk
- the creation of SDK-related rules for the ADuCM3029 platform

File: tools/scripts/build_projects.py [hunk 6]
```diff
@@ -192,14 +192,17 @@ def build(self):
 				new_hdf = True
 		
 		if new_hdf:
-			err = run_cmd(cmd + ' clean_all')
+			err = run_cmd(cmd + ' reset')
```
Comment (buha) [inline tools/scripts/build_projects.py:195]: the changes in build_projects should probably go in the 2nd commit, that's the one that deals with reset and update

File: tools/scripts/stm32.mk [hunk 7]
```diff
@@ -163,8 +163,8 @@ post_build: $(HEX)
 	$(MUTE) $(foreach inc, $(EXTRA_INC_PATHS), sed -i '/Core\/Inc"\/>/a <listOptionValue builtIn="false" value="$(inc)"\/>' $(PROJECT_BUILDROOT)/.cproject;) $(HIDE)
 	$(MUTE) $(foreach flag, $(CPROJECTFLAGS), sed -i '/USE_HAL_DRIVER"\/>/a <listOptionValue builtIn="false" value="$(flag)"\/>' $(PROJECT_BUILDROOT)/.cproject;) $(HIDE)
 
-PHONY += $(PLATFORM)_project_build
-$(PLATFORM)_project_build: $(PROJECT_TARGET) $(LIB_TARGETS)
+PHONY += $(PLATFORM)_sdkbuild
```
Comment (buha) [inline tools/scripts/stm32.mk:168]: commit title suggests this is for Xilinx only but you have a stm32 related change here.

Comment (buha) [reply tools/scripts/stm32.mk:168]: Actually I think you should rather do each rule like you did it in the following commit:
`tools: scripts: *.mk: sdkopen rule `
Put the top-level generic.mk rule altogether with all platform specific rules that it uses in the same commit. That way, when somebody checks-out this commit, the rule will work for any platform.

**Comment by danmois** [comment]
V5

- moved the modifications performed in build_projects.py to the second commit

File: tools/scripts/build_projects.py (hunk 6 above)
Comment (danmois) [reply tools/scripts/build_projects.py:195]: Done

**Comment by danmois** [comment]
V6

- reorganized modifications in commits dealing with specific rules on all applicable targets (as recommended in https://github.com/analogdevicesinc/no-OS/pull/1222#discussion_r795409130)

File: tools/scripts/generic.mk (hunk 4 above)
Comment (adrimbarean) [reply tools/scripts/generic.mk:385]: No need. Whatever we choose it's gonna be great.

**Review by adrimbarean — APPROVED** [review]
Looks good to me.

**Review by buha — APPROVED** [review]
Looks good, but we need to do some testing

File: tools/scripts/aducm.mk [hunk 8]
```diff
@@ -283,4 +283,12 @@ clean_project:
  		-project $(PROJECT_NAME) \
  		-cleanOnly all
 
+PHONY += $(PLATFORM)_sdkclean
+$(PLATFORM)_sdkclean:
+	$(call print,[Delete] SDK created artefacts from $(BUILD_DIR))
```
Comment (buha) [inline tools/scripts/aducm.mk:288]: Could you please delete the "created" word ?
It generates this output and I find it very confusing whether something was created or deleted, even though by careful reading one may understand the proper meaning of it.
```
[iio_demo]$ make sdkclean
[09:52:42] [Delete] SDK created artefacts from /home/dari/Work/no-OS/projects/iio_demo/build/app/Debug
[09:52:42] [Delete] SDK created artefacts from /home/dari/Work/no-OS/projects/iio_demo/build/app/Release
```

File: tools/scripts/stm32.mk [hunk 9]
```diff
@@ -168,6 +170,13 @@ $(PLATFORM)_sdkbuild:
 	$(MUTE) $(STM32CUBEIDE)/$(IDE) -nosplash -application org.eclipse.cdt.managedbuilder.core.headlessbuild \
 		-import $(PROJECT_BUILDROOT) -data $(BUILD_DIR) -build app $(HIDE)
 
+PHONY += $(PLATFORM)_sdkclean
+$(PLATFORM)_sdkclean:
+	$(call print,[Delete] SDK created artefacts from $(DEBUG_DIR))
```
Comment (buha) [inline tools/scripts/stm32.mk:175]: also here, "created"
and everywhere else

**Comment by buha** [comment]
So I've tested this on stm32 and works as expected

**Comment by danmois** [comment]
V7

- Fixed the message displayed by the sdkclean rule (removal of the "created" work)

**Comment by buha** [comment]
remove_dir_action on linux is missing

**Comment by buha** [comment]
let's also implement `make build`

**Comment by danmois** [comment]
V8

- added the build rule
- fixed a bug regarding the sdkbuild for STM32 on Linux (the addition of the remove_dir_action function in generic.mk - required for consistency with the Win implementation)

**Comment by APorumb** [comment]
I've tested building xilinx project on windows machine and works as expected.
